### ISSUE TITLE
Worker pool 

### ISSUE THREAD

**Issue by Btlyons1** [ciscocsirt/netsarlacc#11]
Rework the worker pool design

**Comment by bmenrigh** [comment]
I'm not sure what you have in mind here but workers have a "Work" channel that the put into the WorkQueue channel and then that WorkQueue reads a worker's Work channel and adds work to it.  We didn't make the Work channel buffered though:

`Work        chan WorkRequest`

So writes to a worker's Work channel will block unless the worker is sitting in the select:

`                        select {
                        case work := <-w.Work:`

This blocking will add a tiny bit of latency.  We should make the Work channel have a small buffer.  Only one thing gets added each time the channel comes off the WorkQueue anyways so adding a buffer just stops the synchronous aspect of the channel adding latency.

This would be especially important if we modified the dispatcher to not create a new goroutine every time it adds to the work queue.  I think creating the goroutine adds latency that we don't need either.

**Comment by bmenrigh** [comment]
I did some benchmarking to test whether adding a buffer to the Work chan WorkRequest channel reduces latency.  I'm unable to measure any difference between having it buffered and unbuffered.

I suppose this makes sense because nearly immediately after a worker adds their work channel to the queue, they enter select so they're already waiting by the time any work is added to their channel.

**Comment by bmenrigh** [comment]
I setup netsarlacc on a slow 2-core Intel Atom machine and benchmarked it with Apache Benchmark from a much faster machine.  Both are connected at 1 Gbit.

```
$ ab -q -l -n 50000 -c 50 http://10.155.6.69/test
This is ApacheBench, Version 2.3 <$Revision: 1757674 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 10.155.6.69 (be patient).....done


Server Software:        
Server Hostname:        10.155.6.69
Server Port:            80

Document Path:          /test
Document Length:        Variable

Concurrency Level:      50
Time taken for tests:   42.697 seconds
Complete requests:      50000
Failed requests:        0
Total transferred:      89744453 bytes
HTML transferred:       85444453 bytes
Requests per second:    1171.03 [#/sec] (mean)
Time per request:       42.697 [ms] (mean)
Time per request:       0.854 [ms] (mean, across all concurrent requests)
Transfer rate:          2052.61 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       5
Processing:     1   42   8.4     42     120
Waiting:        1   42   8.4     42     120
Total:          2   43   8.4     42     121

Percentage of the requests served within a certain time (ms)
  50%     42
  66%     44
  75%     45
  80%     46
  90%     52
  95%     56
  98%     63
  99%     69
 100%    121 (longest request)
```

A response rate of ~1170/sec seems to be about the maximum this machine is capable of handling.  Both cores are driven to 100% at that load.

**Comment by bmenrigh** [comment]
I ended up redoing the worker pool design (commit 5476e2705fb77afa20b3bb98753e11e29cb1c8e8) to eliminate the channel-of-channels architecture and the goroutine to take channels off of the channel-of-channels to feed them work is no longer needed.

Benchmarking showed no measurable performance difference between the two.  Considering having just one work channel is simpler it's a net win.

My guess is that in previous versions of Go there was a big performance penalty to having multiple goroutines blocking trying to read a single channel.  If so, that doesn't seem to be the case now.
